geopandas/vectorized.pyx

```diff
@@ -0,0 +1,1029 @@
+    cdef GEOSPreparedPredicate predicate
+    cdef unsigned int n = geoms.size
+
+    cdef np.ndarray[np.uint8_t, ndim=1, cast=True] out = np.empty(n, dtype=np.bool_)
+
+    handle = get_geos_context_handle()
+    other_geom = <GEOSGeometry *> other.__geom__
+
+    # Prepare the geometry if it hasn't already been prepared.
+    # TODO: why can't we do the following instead?
+    #   prepared_geom = GEOSPrepare_r(handle, other_geom)
+    if not isinstance(other, shapely.prepared.PreparedGeometry):
+        other = shapely.prepared.prep(other)
+
+    geos_handle = get_geos_context_handle()
+    prepared_geom = geos_from_prepared(other)
+
+    predicate = get_prepared_predicate(op)
+
+    with nogil:
+        for idx in xrange(n):
+            geom = <GEOSGeometry *> geoms[idx]
+            if geom != NULL:
+                out[idx] = predicate(handle, prepared_geom, geom)
+            else:
+                out[idx] = 0
+
+    return out
+
+cdef GEOSPreparedPredicate get_prepared_predicate(str op) except NULL:
+    if op == 'contains':
+        func = GEOSPreparedContains_r
+    elif op == 'disjoint':
+        func = GEOSPreparedDisjoint_r
+    elif op == 'intersects':
+        func = GEOSPreparedIntersects_r
+    elif op == 'touches':
+        func = GEOSPreparedTouches_r
+    elif op == 'crosses':
+        func = GEOSPreparedCrosses_r
+    elif op == 'within':
+        func = GEOSPreparedWithin_r
+    elif op == 'contains_properly':
+        func = GEOSPreparedContainsProperly_r
+    elif op == 'overlaps':
+        func = GEOSPreparedOverlaps_r
+    elif op == 'covers':
+        func = GEOSPreparedCovers_r
+    elif op == 'covered_by':
+        func = GEOSPreparedCoveredBy_r
+    else:
+        raise NotImplementedError(op)
+
+    return func
+
+
+cdef GEOSPredicate get_predicate(str op) except NULL:
+    cdef GEOSPredicate func
+
+    if op == 'contains':
+        func = GEOSContains_r
+    elif op == 'disjoint':
+        func = GEOSDisjoint_r
+    elif op == 'equals':
+        func = GEOSEquals_r
+    elif op == 'intersects':
+        func = GEOSIntersects_r
+    elif op == 'touches':
+        func = GEOSTouches_r
+    elif op == 'crosses':
+        func = GEOSCrosses_r
+    elif op == 'within':
+        func = GEOSWithin_r
+    elif op == 'overlaps':
+        func = GEOSOverlaps_r
+    elif op == 'covers':
+        func = GEOSCovers_r
+    elif op == 'covered_by':
+        func = GEOSCoveredBy_r
+    else:
+        raise NotImplementedError(op)
+
+    return func
+
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+cpdef vector_binary_predicate(str op,
+                             np.ndarray[np.uintp_t, ndim=1, cast=True] left,
+                             np.ndarray[np.uintp_t, ndim=1, cast=True] right):
+    cdef Py_ssize_t idx
+    cdef GEOSContextHandle_t handle
+    cdef GEOSGeometry *left_geom
+    cdef GEOSGeometry *right_geom
+    cdef GEOSPredicate func
+    cdef unsigned int n = left.size
+
```

Comment:
The calling function does this currently.  We could do it in both places though.  We should test for it regardless.